File: tnrs_api_example.php
<?php

//////////////////////////////////////////////////////
// Submits test data to TNRS API & displays results
//
// * Imports names from test file on local file system
// * Displays input and output at various stages of
// 	 the process.
//////////////////////////////////////////////////////

// Load library of http status codes
require_once("includes/php/status_codes.php");

/////////////////////
// API parameters
/////////////////////

require_once 'server_params.php';	// parameters in ALL_CAPS set here
require_once 'params.php';			// parameters in ALL_CAPS set here

// Path and name of file containing input names and political divisions
$inputfilename = "tnrs_testfile.csv";
$inputfile = $DATADIR.$inputfilename;
$inputfile = "http://bien.nceas.ucsb.edu/bien/wp-content/uploads/2019/07/tnrs_testfile.csv";

// Desired response format
//	Options: json*|xml
// Example: $format="xml";
// NOT YET IMPLEMENTED!
$format="json";

// Number of lines to import
// Use this option to limit test data to small subsample of input file
// Set to number > # of lines in file to import entire file
$lines = 10000000000;
$lines = 4;

// api base url 
$base_url = "https://tnrsapi.xyz/tnrs_api.php";	# Production
$base_url = "http://vegbiendev.nceas.ucsb.edu:8975/tnrs_api.php"; # Dev (port)

/////////////////////////////////////////
// TNRS options
//
// UNDER CONSTRUCTION! - NOT ALL OPTIONS
// CURRENTLY IMPLEMENTED
// 
// Set any option to empty string ("") to 
// use default
// *=default option
/////////////////////////////////////////

// Processing mode
//	Options: resolve*|parse|meta
// 	E.g., $mode="parse"
$mode="resolve";			// Resolve names
//$mode="";					// Same as $mode="resolve";
//$mode="parse";			// Parse names
//$mode="meta";				// Return metadata on TNRS & sources
// $mode="sources";			// List TNRS sources
// $mode="citations";		// Return citations for TNRS & sources
//$mode="classifications";	// Return citations for TNRS & sources

// Taxonomic sources
// One or more of the following, separated by commas, no spaces:
//	tpl,gcc,ildis,gcc,tropicos,usda,ncbi
//	Current default: "tpl,tropicos,usda"
$sources="tropicos,tpl,usda";		

// Classification
// 	Options: tropicos only
// 	E.g., $class="tropicos"
$class="tropicos";

// Matches to return
// 	Options: best*|all
// 	Over-ride command line option -m
$matches="best";
//$matches="all";

// Match accuracy (NOT IMPLEMENTED)
// Must be decimal from 0.05 (default) to 1
//	E.g., $accuracy=0.50;
//	Do not enclose in quotes, except empty string for default
$acc=0.05;

// Constrain by higher taxonomy? (NOT YET IMPLEMENTED)
//	Options: true|false*
//	Boolean: do not enclose in quotes (except can use empty string for default)
$constr_ht=false;

// Constraint by taxonomic source? (NOT YET IMPLEMENTED)
//	Options: true|false*
//	Boolean: do not enclose in quotes (except can use empty string for default)
$constr_ts=false;

/////////////////////////////////////////
// Display options
// 
// * Turn on/off what is echoed to terminal
// * Raw data always displayed
/////////////////////////////////////////

$disp_data_array=false;		// Echo raw data as array
$disp_combined_array=false;	// Echo combined options+data array
$disp_opts_array=false;		// Echo TNRS options as array
$disp_opts=true;			// Echo TNRS options
$disp_json_data=true;		// Echo the options + raw data JSON POST data
$disp_results_json=true;	// Echo results as array
$disp_results_array=false;	// Echo results as array
$disp_results_csv=true;		// Echo results as CSV text
$time=true;					// Echo time elapsed

/////////////////////////////////////////////////////
// Command line options
// Use to over-ride the above parameters
/////////////////////////////////////////////////////

// Get options, set defaults for optional parameters
// Use default if unset
$options = getopt("b:m:");
$batches=isset($options["b"])?$options["b"]:"$NBATCH";	
// $matches=isset($options["m"])?$options["m"]:"$TNRS_DEF_MATCHES";

////////////////////////////////////////////////////////////////
// Main
////////////////////////////////////////////////////////////////

include $timer_on; 	// Start the timer
echo "\n";

///////////////////////////////
// Make options array
///////////////////////////////
$opts_arr = array(
	"sources"=>$sources, 
	"class"=>$class, 
	"mode"=>$mode,
	"acc"=>$acc, 
	"constr_ht"=>$constr_ht, 
	"constr_ts"=>$constr_ts,
	"matches"=>$matches,
	"batches"=>$batches
	);

///////////////////////////////
// Make data array
///////////////////////////////

// Import csv data and convert to array
$data_arr = array_map('str_getcsv', file($inputfile));

# Get subset
$data_arr = array_slice($data_arr, 0, $lines);

if ( $mode=="parse" || $mode=="resolve" || $mode=="" ) {
	// Echo raw data
	echo "The raw data:\r\n";
	foreach($data_arr as $row) {
		foreach($row as $key => $value) echo "$value\t"; echo "\r\n";
	}
	echo "\r\n";

	if ($disp_data_array) {
		echo "The raw data as array:\r\n";
		var_dump($data_arr);
		echo "\r\n";
	}
}

///////////////////////////////
// Merge options and data into 
// json object for post
///////////////////////////////

// Convert to JSON
$json_data = json_encode(array('opts' => $opts_arr, 'data' => $data_arr));	

///////////////////////////////
// Decompose the JSON
// into opt and data
///////////////////////////////

$input_array = json_decode($json_data, true);

if ($disp_combined_array) {
	echo "The combined array:\r\n";
	var_dump($input_array);
	echo "\r\n";
}

$opts = $input_array['opts'];
if ($disp_opts_array) {
	echo "Options array:\r\n";
	var_dump($opts);
	echo "\r\n";
}

if ($disp_opts) {
	// Convert booleans to text for display
	$constr_ht_disp = isset($opts['constr_ht']) ?
		($opts['constr_ht']==true?"true":"false") : "false";
	$constr_ts_disp = isset($opts['constr_ts']) ?
		($opts['constr_ts']==true?"true":"false") : "false";
	$mode_disp = isset($opts['mode']) ? $mode : "resolve";
	
	// Echo the options
	echo "TNRS options:\r\n";
	echo "  sources: " . $opts['sources'] . "\r\n";
	echo "  class: " . $opts['class'] . "\r\n";
	echo "  mode: " . $mode_disp . "\r\n";
	echo "  acc: " . $opts['acc'] . "\r\n";
	echo "  constr_ht: " . $constr_ht_disp . "\r\n";
	echo "  constr_ts: " . $constr_ts_disp . "\r\n";
	echo "  matches: " . $opts['matches'] . "\r\n";
	echo "  batches: " . $opts['batches'] . "\r\n";
	echo "\r\n";
}

if ($disp_json_data) {
	// Echo the final JSON post data
	echo "API input (options + raw data converted to JSON):\r\n";
	echo $json_data . "\r\n\r\n";
}

///////////////////////////////
// Call the API
///////////////////////////////

$url = $base_url;    

// Initialize curl & set options
$ch = curl_init($url);	
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);	// Return response (CRITICAL!)
curl_setopt($ch, CURLOPT_POST, 1);	// POST request
curl_setopt($ch, CURLOPT_POSTFIELDS, $json_data);	// Attach the encoded JSON
curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json')); 

// Send the API call
$response = curl_exec($ch);

// Check status of the response and echo if error
$status = curl_getinfo($ch, CURLINFO_HTTP_CODE);

if ( $status != 201 && $status != 200 ) {
	$status_msg = $http_status_codes[$status];
    die("Error: call to URL $url failed with status $status $status_msg \r\nDetails: $response \r\n");
}

// Close curl
curl_close($ch);

///////////////////////////////
// Echo the results
///////////////////////////////

$results_json = $response;
$results = json_decode($results_json, true);	// Convert JSON results to array

// Echo the JSON response
if ($disp_results_json) {
	echo "API results (JSON)\r\n";
	echo $results_json;
	echo "\r\n\r\n";
}

if ($disp_results_array) {
	echo "API results as array:\r\n";
	var_dump($results);
	echo "\r\n\r\n";
}

if ($disp_results_csv) {
	echo "API results as CSV:\r\n";
	
// 	if ( ! ($mode=="resolve" || $mode=="" || $mode=="parse") ) {
// 		$results = $results[0];	// Remove one level of nesting
// 	}
// 
	foreach ( $results as $rkey => $row ) {
		$rind=array_search( $rkey, array_keys($results) );	# Index: current row
		$cindmax = count( $row )-1;	// Index: last column of current row
	
		if ( $rind==0 ) {
			// Print header
			foreach ( $row as $key => $value )  {	
				$cind=array_search( $key, array_keys($row) );
				$cind==$cindmax?$format="%1s\n":$format="%1s,";
				printf($format, $key);
			}
		}

		// Print data 
		foreach ( $row as $key => $value )  {	
			$cind=array_search( $key, array_keys($row) );
			$cind==$cindmax?$format="%1s\n":$format="%1s,";
			printf($format, $value);
		}
	}
}

///////////////////////////////////
// Echo time elapsed
///////////////////////////////////

include $timer_off;	// Stop the timer
if ($time) echo "\r\nTime elapsed: " . $tsecs . " seconds.\r\n\r\n"; 

?>

File: tnrs_api_example.R
###############################################
# TNRS API Example
###############################################

rm(list=ls())

#################################################
# Parameters & libraries
#################################################

##################
# Base URLs - choose one
##################

# Production
url = "https://tnrsapi.xyz/tnrs_api.php"	

# Public development on vegbiendev (for testing development versions)
url = "http://vegbiendev.nceas.ucsb.edu:9975/tnrs_api.php" 

##################
# Libraries
##################

library(httr)		# API requests
library(jsonlite) # JSON coding/decoding

##################
# Input data
##################

# Example names file 
example_file <- "http://bien.nceas.ucsb.edu/bien/wp-content/uploads/2019/07/tnrs_testfile.csv"

# File of names to resolve, change  as needed. 
# If you use your own file, it must use the same format as the example file: 
# * CSV UTF8
# * first column an integer ID
# * second column the taxon name
names_file <- example_file
	
##################
# Misc parameters
##################

# API variables to clear before each API call
# Avoids spillover between calls
api_vars <- c("mode", "sources", "class", "matches", "acc")

# Response variables to clear before each API call
# Avoids confusion with previous results if API call fails
response_vars <- c("request_json", "response_json", "response")

#################################################
# Functions
#################################################

make_request_json <- function( mode,	# API mode; required
	sources=NULL,		# Taxonomic sources
	class=NULL,			# Family-level classification source
	matches=NULL,	# Matches returned: best|all
	acc=NULL,				# Match accuracy. Float 0-1, default=0.53
	data=NULL 		# Raw data; required if mode %in% c('resolve','parse')
	) {
	######################################
	# Accepts: options parameters and (optionally) data
	# Returns: formatted JSON api request
	######################################
	
	# Set defaults if applicable
	if ( mode=="resolve" || mode=="parse" )  {
		# Convert raw data to JSON
		if (is.null(data) ) stop("ERROR: modes 'resolve' and 'parse' require data!\n")
		data_json <- jsonlite::toJSON(unname(data))
	} 
	
	opts <- data.frame(mode = mode)
	if ( mode=="resolve" ) {
		# Add remaining options if provided
		# If not provided, TNRS will use defaults
		if (!is.null(sources)) opts <- cbind(opts, data.frame(sources = sources))
		if (!is.null(class)) opts <- cbind(opts, data.frame(class = class))
		if (!is.null(matches)) opts <- cbind(opts, data.frame(matches = matches))
		if (!is.null(acc)) opts <- cbind(opts, data.frame(acc = acc))
	}
	opts_json <-  jsonlite::toJSON(opts)
	opts_json <- gsub('\\[','',opts_json)
	opts_json <- gsub('\\]','',opts_json)
	
	# Combine the options and data into single JSON object
	if ( mode=="resolve" || mode=="parse" ) {
		input_json <- paste0('{"opts":', opts_json, ',"data":', data_json, '}' )
	} else {
		input_json <- paste0('{"opts":', opts_json, '}' )
	}
	
	return(input_json)
}

send_request_json <- function( url, request_json ) {
	###################################
	# Accepts: API url + JSON options & data
	# Sends: POST request to url, with JSON
	#		attached to body
	# Returns: JSON response
	###################################
	if ( is.null(url) || is.na(url) ) {
		stop("ERROR: parameter 'url' missing (function send_request_json()'\n")
	}
	if ( is.null(request_json) || is.na(request_json) ) {
		stop("ERROR: parameter 'json_body' missing (function send_request_json()'\n")
	}
	
	response_json <- POST(url = url,
	                  add_headers('Content-Type' = 'application/json'),
	                  add_headers('Accept' = 'application/json'),
	                  add_headers('charset' = 'UTF-8'),
	                  body = request_json,
	                  encode = "json"
	                  )

	return(response_json)
}

decode_response_json <- function( response_json ) {
	###################################
	# Converts resonse json to data frame
	###################################
	
	response_raw <- fromJSON( rawToChar( response_json$content ) ) 
	response <- as.data.frame(response_raw)
	return( response )
}

tnrs_request <- function(url, mode,	# Required
	sources=NULL,		# Taxonomic sources
	class=NULL,			# Family-level classification source
	matches=NULL,	# Matches returned: best|all
	acc=NULL,				# Match accuracy. Float 0-1, default=0.53
	data=NULL 			# Raw data required if mode %in% c('resolve','parse')
	) {
	######################################
	# Accepts: options parameters and (optionally) data
	#		required for TNRS request
	# Sends: POST request to TNRS API
	# Returns: response formatted as data frame
	# 
	# Meta-function which combine functions 
	# make_request_json, send_request_json & 
	# decode_response_json. See component 
	# functions for details
	######################################
	if ( is.null(url) || is.na(url) ) {
		stop("ERROR: parameter 'url' missing (function tnrs_request()'\n")
	}
	if ( is.null(mode) || is.na(mode) ) {
		stop("ERROR: parameter 'mode' missing (function tnrs_request()'\n")
	}

	request_json <- make_request_json(
		mode=mode,
		sources= sources,	
		class= class,
		matches= matches,
		acc= acc,	
		data= data 
		)
	response_json <- send_request_json( url, request_json )
	response_df <- decode_response_json( response_json )
	
	if ( ncol(response_df)==1 ) {
		colnames(response_df) <- "error"
		response_df$http_status <- response_json$status
	}

	return( response_df )
}

specify_decimal <- function(x, k) {
	# Set fixed number of decimals
	if ( is.na(x) || is.null(x) ) {
		x.formatted <- x
	} else {
		x.formatted <- format(round(x, k), nsmall=k)
	}
	return( x.formatted )
}

#################################################
# Main
#################################################

#################################
# Example 1: Metadata
#
# Available metadata calls: 
# "meta", "sources", "class", "citations", "collaborators"
# print(response) to see the complete results of each call
#################################

# TNRS application version and database version
mode <- "meta"		
rm( list = Filter( exists, response_vars ) )
response <- tnrs_request(url=url, mode=mode)
db_version  <- response$db_version
db_date   <- response$build_date
if( "app_version" %in% colnames(response) ) {
	tnrs_version  <- response$app_version
} else {
	tnrs_version  <- response$code_version
}

# Available sources
mode <- "sources"		
rm( list = Filter( exists, response_vars ) )
response <- tnrs_request(url=url, mode=mode)
sources  <- as.vector(t(response$sourceName))
source.details <- response[ , c(
	"sourceID", "sourceName", "sourceNameFull", "version", 
	"sourceReleaseDate", "tnrsDateAccessed"
	)]

# Available classifications
mode <- "classifications"		
rm( list = Filter( exists, response_vars ) )
response <- tnrs_request(url=url, mode=mode)
classifications  <- as.vector(t(response$sourceName))

# TNRS and source citations
mode <- "citations"		
rm( list = Filter( exists, response_vars ) )
citations <- tnrs_request(url=url, mode=mode)
wfo_citation <- citations$citation[ citations$source=="wfo" ]
cact_citation <- citations$citation[ citations$source=="cact" ]

# TNRS collaborators
mode <- "collaborators"		
rm( list = Filter( exists, response_vars ) )
collaborators <- tnrs_request(url=url, mode=mode)
collaborator.codes  <- as.vector(t(collaborators$collaboratorName))

# Display results
cat("TNRS version: ", tnrs_version, "\n")
cat("DB version: ", db_version, " (", db_date, ")\n")
cat("Available taxonomic sources: ", sources, "\n")
cat("Available family classifications: ", classifications, "\n")
cat(" \r\n")

cat("Taxonomic source details:\n")
print(source.details, row.names=FALSE)
cat(" \r\n")

cat("\r\nCitations:")
for (i in 1:nrow(citations)) { 
	cat(citations$source[ i ], ":\r\n", sep="")
	cat("  ", citations$citation[ i ], ":\r\n", sep="")
}
	cat(" \r\n")

cat("\r\nTNRS collaborators:\r\n")
cat(collaborator.codes, sep = "\r\n")
cat(" \r\n")

#################################
# Example 2: Resolve mode, best match only
#################################

# Load test data
data <- read.csv(names_file, header=FALSE)
colnames(data) <- c("ID", "Name_submitted")
data <- head(data, 10) # Just a sample
cat("Raw names:\n")
print(data)

# Clear existing variables
suppressWarnings( rm( list = Filter( exists, c(response_vars, api_vars ) ) ) )

# Set options
sources <- "wcvp,wfo"		# Taxonomic sources
class <- "wfo"					# Family classification. Only current option: "wfo"
mode <- "resolve"			# Processing mode
matches <- "best"			# Return best match only

response <- tnrs_request(url=url, mode=mode, matches=matches, 
	source=sources, class=class, data=data	)
if ( colnames(response)[1]=="error" ) {
	print( response )
} else {
	# Format overall_score to 2 decimals
	response$Overall_score <- as.numeric(lapply( 
		as.numeric(response$Overall_score), FUN=specify_decimal, k=2
		))
	results_cols_basic <- c(
		"ID", "Name_submitted","Overall_score", 	"Name_matched", 
		"Accepted_name", "Accepted_name_author",	"Source"
	)
	print( response[ , results_cols_basic ]	)
}

#################################
# Example 3: Resolve mode, all matches
#################################

# Clear existing variables
suppressWarnings( rm( list = Filter( exists, c(response_vars, api_vars ) ) ) )

# Fewer rows of the shorter names
data.small <- data[c(1:2, 4:7), ]

# Set options
mode <- "resolve"						
sources <- "wfo,wcvp"	
class <- "wfo"
matches <- "all"		# Return all matches

response <- tnrs_request(url=url, mode=mode, matches=matches, 
	source=sources, class=class, data= data.small	)
if ( colnames(response)[1]=="error" ) {
	print( response )
} else {
	response$Overall_score <- as.numeric(lapply( 
		as.numeric(response$Overall_score), FUN=specify_decimal, k=2
		))
	results_cols <- c(results_cols_basic,
		 c("Overall_score_order", "Highertaxa_score_order", "Warnings")
	)
	print( response[ , results_cols ]	)
}

#################################
# Example 3: Resolve mode, all matches, 
# Custom match threshold sets minimum
# value of Overall_score
#################################

# Clear existing variables
suppressWarnings( rm( list = Filter( exists, c(response_vars, api_vars ) ) ) )

# Set options
mode <- "resolve"						
sources <- "wfo,wcvp"	
class <- "wfo"
matches <- "all"	
acc <- 0.9					# High match threshold

response <- tnrs_request(url=url, mode=mode, matches=matches, 
	source=sources, class=class, acc=acc, data= data.small	)
if ( colnames(response)[1]=="error" ) {
	print( response )
} else {
	response$Overall_score <- as.numeric(lapply( 
		as.numeric(response$Overall_score), FUN=specify_decimal, k=2
		))
	results_cols <- c(results_cols_basic,
		 c("Overall_score_order", "Highertaxa_score_order", "Warnings")
	)
	print( response[ , results_cols ]	)
}

#################################
# Example 5: Parse mode
#################################

# Clear existing variables
suppressWarnings( rm( list = Filter( exists, c(response_vars, api_vars ) ) ) )

# Set options
mode <- "parse"			# Processing mode

response <- tnrs_request(url=url, mode=mode, matches=matches, 
	source=sources, class=class, data=data	)
if ( colnames(response)[1]=="error" ) {
	print( response )
} else {
	results_cols <- c(
		"ID", "Name_submitted", "Family", "Genus", "Specific_epithet", "Author"
	)
	print( response[ , results_cols ]	)
}

#################################
# Example 5: TNRS data dictionary
#################################

# Available classifications
mode <- "dd"		
rm( list = Filter( exists, response_vars ) )
dd <- tnrs_request(url=url, mode=mode)
dd
